### PR TITLE
Port bug fixes and changes noticed while testing COS integration in k8s charm

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
There were a number of issues encountered while testing the COS integration in the k8s charm. The fixes for these issues are still missing from this repo

## Solution
Port fixes

### CODE REVIEW

**Review by taurus-forever — APPROVED** [review]

File: src/relations/cos.py [hunk 1]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
...
         )
 
+    def is_relation_cos_related(self, event) -> bool:
+        """Whether relation is related to the metrics endpoint."""
+        if not self.relation_exists:
+            return False
+
+        return (
+            hasattr(event, "relation")
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py:124]: i believe we are checking if relation exists in a condition right above this return statement

File: src/snap.py [hunk 2]
```diff
@@ -184,7 +184,14 @@ def update_mysql_router_service(self, *, enabled: bool, tls: bool = None) -> Non
...
-        self, *, enabled: bool, config: "relations.cos.ExporterConfig" = None
+        self,
+        *,
+        enabled: bool,
+        config: "relations.cos.ExporterConfig" = None,
+        tls: bool = None,
+        key: str = None,
+        certificate: str = None,
+        certificate_authority: str = None,
     ) -> None:
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/snap.py:195]: these are only k8s args at the moment as only k8s supports tls currently. in the near future, we will support tcp connections to mysqlrouter vm and thus will implement TLS (where these args will be used)

File: src/relations/cos.py [hunk 3]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
...
+        logger.debug("Waiting until router HTTP server authenticates")
+        try:
+            for attempt in tenacity.Retrying(
+                reraise=True,
+                stop=tenacity.stop_after_delay(30),
+                wait=tenacity.wait_fixed(5),
+            ):
+                with attempt:
+                    # do not verify tls certs as default certs do not have 127.0.0.1
+                    # in its list of IP SANs
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py:138]: i dont believe so, at least not with the default autogenerated TLS certificate used by Router's HTTP server. Below is the summary of the cert. I think key is that there are no SANs and constraint CA=False

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = MySQL_Router_Auto_Generated_CA_Certificate
        Validity
            Not Before: Mar 25 10:03:21 2024 GMT
            Not After : Mar 23 10:03:21 2034 GMT
        Subject: CN = MySQL_Router_Auto_Generated_Router_Certificate
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b6:d7:a3:42:70:9c:b6:1e:25:ee:ae:62:a6:82:
                    48:85:ff:0a:30:96:79:44:98:0f:f4:5a:86:d1:71:
                    27:ec:7b:f9:fa:88:38:87:d3:29:2e:1c:cb:f1:47:
                    a0:5c:83:79:fb:69:ae:db:03:3e:2b:e3:09:62:44:
                    36:cc:76:4c:b2:36:7b:bf:19:e3:bf:23:bc:79:1c:
                    e0:8e:cf:0a:1d:9a:ef:3f:f3:9d:d2:31:09:ec:7a:
                    c1:80:19:42:6f:11:cc:79:75:ad:96:cb:2a:60:7a:
                    b5:8e:39:6b:b2:a8:9d:fb:95:35:38:59:c1:55:c8:
                    2e:24:4f:5a:be:46:cd:2d:26:79:4e:0d:28:ab:77:
                    a8:2e:b3:9a:45:89:af:43:f8:58:43:26:b9:18:d3:
                    f3:b1:c9:f3:61:ff:b7:d8:1c:53:79:78:a5:e0:61:
                    2e:e5:ac:63:3d:19:da:bd:8b:0e:79:38:db:86:11:
                    60:5b:c8:38:05:d3:b3:0d:0f:61:ce:98:b2:74:bd:
                    4c:a7:d2:35:19:59:ec:45:95:27:df:6f:5c:28:59:
                    bf:ff:5a:7d:b5:6e:33:c3:6a:32:88:f2:32:f5:e5:
                    86:a5:5d:fe:5a:9d:8e:50:09:55:a8:14:2b:b1:af:
                    77:1a:7c:2d:d2:4a:89:85:97:6b:80:ca:52:84:1d:
                    e1:ad
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        5b:1a:3a:24:80:fc:78:a5:03:95:02:32:84:64:de:d7:8e:6c:
        a5:b4:47:fd:84:7a:3b:d9:4a:0f:2e:7f:04:67:6f:e3:e7:09:
        f3:d3:6c:5a:25:d8:94:94:37:88:fb:41:b6:61:9f:1c:21:53:
        a4:3b:b7:9b:e8:c7:55:cf:86:49:6a:12:2e:a5:16:72:33:71:
        1a:75:e4:53:00:8a:b1:86:6a:fe:50:9b:c2:9c:2e:0b:34:a9:
        77:bb:07:cc:7d:89:d6:89:e7:ae:87:59:1a:a3:91:4d:a3:5a:
        71:4e:4d:56:95:c4:8e:17:c9:7b:a1:d6:b7:b0:15:4b:76:b9:
        26:5b:79:c9:28:af:bd:35:df:01:37:6b:e4:f2:b1:05:33:92:
        b2:90:4c:9c:02:74:e1:c5:27:ab:ea:d1:6d:5a:80:9f:df:04:
        2f:d1:ec:d8:79:21:d9:85:15:96:12:df:e1:a9:6f:cd:49:f4:
        44:8b:20:5d:24:a2:98:4c:41:a7:35:c6:d4:47:4c:b1:a2:53:
        da:7c:7c:6c:f3:f7:d1:45:4a:02:f5:19:80:57:28:98:c1:eb:
        8d:a9:90:54:62:05:5c:95:5a:ff:3f:25:d0:dd:ea:22:6b:e2:
        04:19:08:b7:4c:1d:d7:36:6c:1b:03:3d:b6:d3:48:14:02:e9:
        e5:99:e4:37
```

File: src/abstract_charm.py [hunk 4]
```diff
@@ -84,9 +85,9 @@ def _logrotate(self) -> logrotate.LogRotate:
         """logrotate"""
 
     @property
-    @abc.abstractmethod
     def _cos(self) -> relations.cos.COSRelation:
         """COS"""
+        return self._cos_relation
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/abstract_charm.py]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/abstract_charm.py [hunk 5]
```diff
@@ -116,12 +117,19 @@ def _tls_certificate(self) -> str:
...
+
     def _cos_exporter_config(self, event) -> typing.Optional[relations.cos.ExporterConfig]:
         """Returns the exporter config for MySQLRouter exporter if cos relation exists"""
         cos_relation_exists = self._cos.relation_exists and not self._cos.is_relation_breaking(
             event
         )
-        return self._cos.exporter_user_config if cos_relation_exists else None
+        if cos_relation_exists:
+            return self._cos.exporter_user_config
+        return None
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/abstract_charm.py]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/container.py [hunk 6]
```diff
@@ -134,13 +134,24 @@ def update_mysql_router_service(self, *, enabled: bool, tls: bool = None) -> Non
...
     def update_mysql_router_exporter_service(
-        self, *, enabled: bool, config: "relations.cos.ExporterConfig" = None
+        self,
+        *,
+        enabled: bool,
+        config: "relations.cos.ExporterConfig" = None,
+        tls: bool = None,
+        key: str = None,
+        certificate: str = None,
+        certificate_authority: str = None,
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/container.py:144]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50. now validating these values when tls=True

File: src/relations/cos.py [hunk 7]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
...
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
         )
 
+    def is_relation_cos_related(self, event) -> bool:
+        """Whether relation is related to the metrics endpoint."""
+        if not self.relation_exists:
+            return False
+
+        return (
+            hasattr(event, "relation")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/relations/cos.py [hunk 8]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
 
         return (
             isinstance(event, ops.RelationBrokenEvent)
-            and event.relation.id == self.charm.model.relations[self._NAME][0].id
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
         )
 
+    def is_relation_cos_related(self, event) -> bool:
+        """Whether relation is related to the metrics endpoint."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py]: agreed, it should be `whether relation is related to the cos_agent endpoint.` however, this method was removed in 621c276a7c23c306f68f9078a73da9f4ff47bc50 as part of the requested changes

File: src/relations/cos.py [hunk 9]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
...
+        return (
+            hasattr(event, "relation")
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
+        )
+
+    def _wait_until_http_server_authenticates(self) -> None:
+        """Wait until the router HTTP server authenticates with the monitoring credentials."""
+        logger.debug("Waiting until router HTTP server authenticates")
+        try:
+            for attempt in tenacity.Retrying(
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py:121]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/relations/cos.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/relations/cos.py [hunk 10]
```diff
@@ -109,16 +111,52 @@ def is_relation_breaking(self, event) -> bool:
...
+            ):
+                with attempt:
+                    # do not verify tls certs as default certs do not have 127.0.0.1
+                    # in its list of IP SANs
+                    response = requests.get(
+                        f"https://127.0.0.1:{self._HTTP_SERVER_PORT}/api/20190715/routes",
+                        auth=(self._MONITORING_USERNAME, self._get_monitoring_password()),
+                        verify=False,
+                    )
+                    assert response.status_code == 200
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/workload.py [hunk 11]
```diff
@@ -124,19 +124,21 @@ def _disable_tls(self) -> None:
...
         exporter_config: "relations.cos.ExporterConfig",
+        event_is_cos_related: bool,
         key: str = None,
         certificate: str = None,
+        certificate_authority: str = None,
     ) -> None:
         """Reconcile all workloads (router, exporter, tls)."""
-        if tls and not (key and certificate):
+        if tls and not (key and certificate and certificate_authority):
             raise ValueError("`key` and `certificate` arguments required when tls=True")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/workload.py:141]: addressed in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/workload.py [hunk 12]
```diff
@@ -124,19 +124,21 @@ def _disable_tls(self) -> None:
...
-        logger.debug("Deleting TLS files")
+        logger.debug("Deleted TLS files")
 
     def reconcile(
         self,
         *,
         tls: bool,
         unit_name: str,
         exporter_config: "relations.cos.ExporterConfig",
+        event_is_cos_related: bool,
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/workload.py]: good point! worked myself into knots when debugging something and thought this was necessary. removed kwarg `event_is_cos_related` in 621c276a7c23c306f68f9078a73da9f4ff47bc50

File: src/relations/cos.py [hunk 13]
```diff
@@ -109,16 +111,42 @@ def is_relation_breaking(self, event) -> bool:
...
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
         )
 
+    def _wait_until_http_server_authenticates(self) -> None:
+        """Wait until the router HTTP server authenticates with the monitoring credentials."""
+        logger.debug("Waiting until router HTTP server authenticates")
+        try:
+            for attempt in tenacity.Retrying(
+                retry=tenacity.retry_if_exception_type(AssertionError)
+                | tenacity.retry_if_exception_type(requests.exceptions.HTTPError),
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py:123]: yes, still valid syntax on python 3.8:

```
>>> import tenacity
>>> for attempt in tenacity.Retrying(retry=tenacity.retry_if_exception_type(AssertionError) | tenacity.retry_if_exception_type(ValueError), reraise=True, stop=tenacity.stop_after_delay(30), wait=tenacity.wait_fixed(5)):
...     with attempt:
...         if attempt.retry_state.attempt_number % 2:
...             print("assertion error")
...             assert False
...         else:
...             print("value error")
...             raise ValueError("test")
... 
assertion error
value error
assertion error
value error
assertion error
value error
assertion error
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/tenacity/__init__.py", line 347, in __iter__
    do = self.iter(retry_state=retry_state)
  File "/usr/local/lib/python3.8/site-packages/tenacity/__init__.py", line 325, in iter
    raise retry_exc.reraise()
  File "/usr/local/lib/python3.8/site-packages/tenacity/__init__.py", line 158, in reraise
    raise self.last_attempt.result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "<stdin>", line 5, in <module>
AssertionError
>>> print(sys.version_info)
sys.version_info(major=3, minor=8, micro=19, releaselevel='final', serial=0)
```

File: src/relations/cos.py [hunk 14]
```diff
@@ -109,16 +111,42 @@ def is_relation_breaking(self, event) -> bool:
 
         return (
             isinstance(event, ops.RelationBrokenEvent)
-            and event.relation.id == self.charm.model.relations[self._NAME][0].id
+            and event.relation.id == self._charm.model.relations[self._NAME][0].id
         )
 
+    def _wait_until_http_server_authenticates(self) -> None:
+        """Wait until the router HTTP server authenticates with the monitoring credentials."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/cos.py]: improved wording in f06b76a5ee62ac1d7b7528ecb2d9a373d85b21d1

**Review by paulomach — COMMENTED** [review]
Maybe I am sleepy, but I did not manage to make it work @shayancanonical ping me tomorrow please.

Code looks good

**Review by paulomach — APPROVED** [review]
Working nicely

**Pull request merged**